### PR TITLE
Make `torch.det()` support complex input.

### PR DESCRIPTION
**Pull request by nikitaved**
As per title. A minor fix required to make it available for the CPU (`fmod` does not support complex).
For CUDA requires [#45898 ](https://github.com/pytorch/pytorch/pull/45898).

### CODE REVIEW

**Comment by nikitaved** [comment]
Once [#45898 ](https://github.com/pytorch/pytorch/pull/45898) is merged, I plan to extend the existing tests to complex types.

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -1146,6 +1147,8 @@ def method_tests():
             NO_ARGS, 'symmetric_psd', (), NO_ARGS, [skipCPUIfNoLapack, skipCUDAIfNoMagma]),
         ('det', lambda: random_symmetric_pd_matrix(S),
             NO_ARGS, 'symmetric_pd', (), NO_ARGS, [skipCPUIfNoLapack, skipCUDAIfNoMagma]),
+        ('det', lambda: random_hermitian_matrix(S),
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:1150]: this test is running for `torch.double` because:
1. `random_hermitian_matrix` returns a double tensor (which I think we should update to `torch.cdouble`) https://github.com/pytorch/pytorch/blob/298e0e0d573a90ed516f32051151c32e9b309c0f/torch/testing/_internal/common_utils.py#L1507
2. `det` doesn't currently exist in this list https://github.com/pytorch/pytorch/blob/298e0e0d573a90ed516f32051151c32e9b309c0f/tools/autograd/gen_variable_type.py#L156, so when backward is run with complex input, it would throw a `RuntimeError`.
3. To actually run a method_test for complex, you would have to have to add 'complex' to the substring so in this case, it would be `complex_hermitian' and also add the name of the test to this list: https://github.com/pytorch/pytorch/blob/298e0e0d573a90ed516f32051151c32e9b309c0f/test/test_autograd.py#L4825

File: test/test_torch.py [hunk 2]
```diff
@@ -7290,6 +7290,55 @@ def run_test(p):
         with self.assertRaisesRegex(RuntimeError, "chain_matmul: Expected one or more matrices"):
             torch.chain_matmul()
 
+    # no CUDA LU yet, used for torch.det())
+    @onlyCPU
+    @skipCPUIfNoLapack
```
Comment (anjali411) [inline test/test_torch.py:7295]: @mruberry should this test go in `test_unary_ops` or `test_linalg`?

**Review by anjali411 — CHANGES_REQUESTED** [review]

File: test/test_torch.py (hunk 2 above)
Comment (mruberry) [reply test/test_torch.py:7295]: test_linalg

Comment (anjali411) [reply test/test_torch.py:7295]: alright then @nikitaved we should just update this test: https://github.com/pytorch/pytorch/blob/8e8fb8542e806b4c966c6c04bc847f915abac61f/test/test_linalg.py#L162

**Review by anjali411 — CHANGES_REQUESTED** [review]
@nikitaved this looks good overall. could you update the autograd test for `det` to run for complex? also let's just extend the test for det in `test_linalg.py`

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 3]
```diff
@@ -28,7 +28,8 @@ static inline std::tuple<Tensor, Tensor> _lu_det_P_diag_U(const Tensor& self) {
   std::tie(lu, pivs, infos) = at::_lu_with_info(self, /*pivot=*/true, /*check_errors=*/false);
   TORCH_CHECK(infos.ge(0).all().item<uint8_t>(), "Invalid argument passed to lu");
   auto n = self.size(-1);
-  auto num_exchanges = (at::arange(1, n + 1, pivs.options()) != pivs).sum(-1, /*keepdim=*/false, /*dtype=*/self.scalar_type()).fmod_(2);
+  auto num_exchanges = (at::arange(1, n + 1, pivs.options().dtype(at::kLong)) != pivs.to(at::kLong))
```
Comment (ngimel) [inline aten/src/ATen/native/LinearAlgebra.cpp:31]: are you converting to kLong to avoid possible overflows? Even in that case setting dtype=at::kLong in `sum` should be enough

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 4]
```diff
@@ -28,7 +28,8 @@ static inline std::tuple<Tensor, Tensor> _lu_det_P_diag_U(const Tensor& self) {
...
+    .sum(-1, /*keepdim=*/false, /*dtype=*/at::kLong).fmod_(2);
   // NB: the `.contiguous()` call is added due to the bug in `.prod()` as reported in
   // issue #https://github.com/pytorch/pytorch/issues/34061
```
Comment (ngimel) [inline aten/src/ATen/native/LinearAlgebra.cpp:34]: #34061 looks to be fixed, should we remove this workaround?

File: test/test_torch.py [hunk 5]
```diff
@@ -7290,6 +7290,55 @@ def run_test(p):
...
+    @dtypes(torch.complex64, torch.complex128)
+    def test_det_complex_only(self, device, dtype):
+        scalar_dtype = {
+            torch.complex64: torch.float32,
+            torch.complex128: torch.float64
+        }
+
+        def run_test(*sizes):
+            x = torch.rand(*sizes, device=device, dtype=dtype)
+            x = x.view(-1, sizes[-2], sizes[-1])
```
Comment (ngimel) [inline test/test_torch.py:7305]: why are you sending 4d sizes if you immediately turn them into 3d?

File: test/test_torch.py [hunk 6]
```diff
@@ -7290,6 +7290,55 @@ def run_test(p):
...
+
+            if TEST_NUMPY:
+                x_detached = x.detach()
```
Comment (ngimel) [inline test/test_torch.py:7308]: x does not require grad, so you don't need detach here?

File: test/test_torch.py [hunk 7]
```diff
@@ -7290,6 +7290,55 @@ def run_test(p):
...
+                for i in range(x.size(0)):
+                    det_numpy = np.linalg.det(x_detached.select(0, i).cpu().numpy())
+                    det_pytorch = x.select(0, i).det()
+                    self.assertEqual(
+                        det_pytorch, det_pytorch.new_tensor(det_numpy),
```
Comment (ngimel) [inline test/test_torch.py:7313]: `torch.from_numpy(det_numpy).to(det_pytorch.device)`, we are trying to not use `new_tensor` as it is an old style API. Alternatively, you can use `compare_with_numpy` here.

File: test/test_torch.py (hunk 5 above)
Comment (nikitaved) [reply test/test_torch.py:7305]: I only create a view for easier comparison against NumPy per each matrix in the batch.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 3 above)
Comment (nikitaved) [reply aten/src/ATen/native/LinearAlgebra.cpp:31]: `n` is also `long` I suppose.

Comment (nikitaved) [reply aten/src/ATen/native/LinearAlgebra.cpp:31]: Maybe actually a better fix would be to return `pivs` as integers, as they are always integers.

Comment (nikitaved) [reply aten/src/ATen/native/LinearAlgebra.cpp:31]: Hmm, now `pivs` are actually integers, from master:
```python
In [1]: import torch

In [2]: x = torch.rand(3, 3, dtype=torch.cdouble)

In [3]: x
Out[3]: 
tensor([[0.3119+0.0148j, 0.6229+0.6629j, 0.0268+0.5217j],
        [0.7130+0.8608j, 0.8461+0.4621j, 0.3453+0.4870j],
        [0.5092+0.6666j, 0.7671+0.6573j, 0.1088+0.0076j]],
       dtype=torch.complex128)

In [4]: x.lu()
Out[4]: 
(tensor([[ 0.7130+0.8608j,  0.8461+0.4621j,  0.3453+0.4870j],
         [ 0.1882-0.2065j,  0.3683+0.7506j, -0.1388+0.5013j],
         [ 0.7500+0.0296j,  0.3839-0.0066j, -0.0858-0.5612j]],
        dtype=torch.complex128),
 tensor([2, 2, 3], dtype=torch.int32))

```

Comment (nikitaved) [reply aten/src/ATen/native/LinearAlgebra.cpp:31]: Sorry, just changing the `dtype` in `sum` is indeed sufficient :)

**Comment by nikitaved** [comment]
Thank you very much for your input! I updated the PR. The tests from `common_methods_invocations.py` are disabled as backward for SVD is not yet implemented for complex.

@anjali411 , could you please tell whether there is still a limitation of autograd not being able to process functions with several inputs of complex type?

File: aten/src/ATen/native/cuda/ReduceSumProdKernel.cu [hunk 8]
```diff
@@ -88,7 +88,7 @@ static void nansum_kernel_cuda(TensorIterator& iter) {
 
 static void prod_kernel_cuda(TensorIterator& iter) {
   auto general_dispatcher = [](TensorIterator& iter) {
-    AT_DISPATCH_ALL_TYPES(iter.dtype(), "prod_cuda", [&]() {
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX(iter.dtype(), "prod_cuda", [&]() {
       prod_functor<scalar_t>{}(iter);
     });
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/ReduceSumProdKernel.cu:93]: Add tests for complex prod.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py]: Thank you! Unfortunately, det_backward relies on svd_backward, and the latter does not yet work with complex input.

File: aten/src/ATen/native/cuda/ReduceSumProdKernel.cu (hunk 8 above)
Comment (nikitaved) [reply aten/src/ATen/native/cuda/ReduceSumProdKernel.cu:93]: NOTE: backward for prod will not work for complex inputs, because it relies on `nonzero`, which is not yet implemented for the complex numbers.

Comment (anjali411) [reply aten/src/ATen/native/cuda/ReduceSumProdKernel.cu:93]: yes this sounds good let's update the backward formula in the next PR. This PR  https://github.com/pytorch/pytorch/pull/46596 enables backward for nonzero for complex. In general, to enable backward for an op for complex, you would need to add an entry in `GRADIENT_IMPLEMENTED_FOR_COMPLEX` in  `gen_variable.py`.

File: test/test_torch.py (hunk 2 above)
Comment (nikitaved) [reply test/test_torch.py:7329]: Hmm, when I try to move things into `test_linalg.py`, some checks fail, unlike for `test_torch.py`.

Comment (nikitaved) [reply test/test_torch.py]: All fine, the tests are for quite large tensors, so we run into overflow issues.

**Comment by anjali411** [comment]
> could you please tell whether there is still a limitation of autograd not being able to process functions with several outputs of complex type?

No I don't think so. Please let me know if you are running into any issues though

**Review by anjali411 — APPROVED** [review]
lgtm

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 4 above)
Comment (anjali411) [reply aten/src/ATen/native/LinearAlgebra.cpp:35]: to @ngimel 's point, let's remove this workaround since #34061 is fixed

**Comment by anjali411** [comment]
@nikitaved can you sign the CLA, rebase the PR and remove the `.contiguous()` workaround please?

**Comment by nikitaved** [comment]
@anjali411 , sorry, force-pushed from the wrong machine, now it should be fixed.

**Comment by anjali411** [comment]
@nikitaved would you like to create a follow-up PR to add complex autograd support for `torch.det` and `torch.prod`?

**Comment by nikitaved** [comment]
@anjali411 , turns out `torch.det` is a rabbit hole, its backward does depend on `cumsum, cumprod, index_select, index_add` and counting. Since all these methods have quite an extensive test coverage, I will submit one PR per each.

**Comment by mruberry** [comment]
Hey @nikitaved! @anjali411 is actually on pto at the moment, but that plan makes a lot of sense. 

I added complex autograd support for torch.linalg.det as a task in the linear algebra tacking issue (https://github.com/pytorch/pytorch/issues/42666) with a link to your comment. Let's track the work there.

**Comment by nikitaved** [comment]
@mruberry , thank you, will update the issue there. Meanwhile, can I request your reviews on these PRs?

**Comment by nikitaved** [comment]
It turns out there is circular relationship between dependencies, so, it is going to be a single PR.

**Comment by mruberry** [comment]
> @mruberry , thank you, will update the issue there. Meanwhile, can I request your reviews on these PRs?

Of course.
